pkg/collector/corechecks/servicediscovery/module/envs.go

```diff
@@ -153,3 +184,135 @@ func getEnvs(proc *process.Process) (map[string]string, error) {
 	}
 	return envs, nil
 }
+
+// EnvScanner reads the environment variables from /ptoc file.
+// It collects only those variables that match the target map if the map is not empty,
+// otherwise collect all environment variables.
+type EnvScanner struct {
+	file    *os.File          // open pointer to environment variables file
+	scanner *bufio.Scanner    // scanner to read file data
+	targets map[uint64]string // map of environment variables of interest
+	envs    map[string]string // collected environment variables
+}
+
+// hashBytes return hash value of a bytes array using FNV-1a hash function
+func hashBytes(b []byte) uint64 {
+	h := fnv.New64a()
+	h.Write(b)
+	return h.Sum64()
+}
+
+func buildTargetsMap() map[uint64]string {
+	targetsMap := make(map[uint64]string)
+
+	for _, target := range targetEnvs {
+		targetsMap[hashBytes([]byte(target))] = target
+	}
+	for _, target := range apm.JavaDetectorEnvs {
+		targetsMap[hashBytes([]byte(target))] = target
+	}
+	for _, target := range apm.DotNetDetectorEnvs {
+		targetsMap[hashBytes([]byte(target))] = target
+	}
+	return targetsMap
+}
+
+// nullTerminatedString used to tokenize the scanner data
+func nullTerminatedString(data []byte, atEOF bool) (advance int, token []byte, err error) {
+	if i := bytes.IndexByte(data, 0); i >= 0 {
+		return i + 1, data[0:i], nil
+	}
+	if atEOF && len(data) > 0 {
+		// end of the file, return the remaining data
+		return len(data), data, nil
+	}
+
+	return 0, nil, nil
```

Comment:
Is this implementation correct? Why does zeroSplitter in pkg/security/ptrace have different return values here? Also, can't we just reuse NewTextScannerIterator() which is exported?

Reply:
I reverted `ptracer` package import due to impact, I copied function `zeroSplitter` from `ptracer` package